### PR TITLE
WIP: Xorg user session

### PR DESCRIPTION
**Pull request by plfiorini**

### CODE REVIEW

**Comment by plfiorini** [comment]
Status: Xorg starts as a user, the logs in ~sddm/.local/share/xorg/Xorg.0.log reports that it has started with logind but for some reason the devices are paused in fact I can't use the keyboard and even if the greeter is started I can't see anything.

@davidedmundson can you give me a hand to figure out what's going on here?

File: src/daemon/VirtualTerminal.cpp [hunk 1]
```diff
@@ -102,6 +102,11 @@ namespace SDDM {
                 qDebug() << "VT mode didn't need to be fixed";
         }
 
+        int openVt(int vt) {
```
Comment (plfiorini) [inline src/daemon/VirtualTerminal.cpp]: maybe rename this to openTty

File: src/daemon/XorgDisplayServer.cpp [hunk 2]
```diff
@@ -117,94 +116,89 @@ namespace SDDM {
...
-
-            //create pipe for communicating with X server
-            //0 == read from X, 1== write to from X
-            int pipeFds[2];
-            if (pipe(pipeFds) != 0) {
-                qCritical("Could not create pipe to start X server");
-            }
+            // jump to vt
+            m_ttyFd = VirtualTerminal::openVt(displayPtr()->terminalId());
+            VirtualTerminal::jumpToVt(displayPtr()->terminalId(), m_ttyFd);
```
Comment (plfiorini) [inline src/daemon/XorgDisplayServer.cpp]: jump to vt after pam auth?

File: src/daemon/XorgDisplayServer.cpp [hunk 3]
```diff
@@ -302,6 +274,75 @@ namespace SDDM {
         mainConfig.load();
     }
 
+    void XorgDisplayServer::onRequestChanged() {
+        m_auth->request()->setFinishAutomatically(true);
+    }
+
+    void XorgDisplayServer::onSessionStarted(bool success) {
+m_display = QLatin1String(":0");
```
Comment (plfiorini) [inline src/daemon/XorgDisplayServer.cpp]: need to receive the display from helper via socket

File: src/daemon/XorgDisplayServer.cpp [hunk 4]
```diff
@@ -117,94 +116,89 @@ namespace SDDM {
...
                 // return fail
                 return false;
             }
+
+            // generate auth file
+            addCookie(m_authPath);
+            changeOwner(m_authPath);
+
+            qDebug() << "Xorg started";
+            m_started = true;
```
Comment (plfiorini) [inline src/daemon/XorgDisplayServer.cpp]: need to set m_display for test mode here

**Comment by davidedmundson** [comment]
Found what's wrong.

X starts goes through PAM and gets its own logind session.
The greeter starts, goes through PAM and gets its own logind session.

The session with X then gets the pauseDevices signal from logind.
This stops the devices working

(weirdly no attempt to activate later sends the resumeDevice signal)

Solution will be to spawn the greeter from the same sddm-helper as the display manager.
(or conversely to start X from the same sddm-helper as the greeter)

Which seems to be what GDM does

**Comment by plfiorini** [comment]
@davidedmundson thanks! it should be fairly easy to implement it. will get back to work asap

**Comment by plfiorini** [comment]
Note to self: Xorg user session might break existing workflows such as Xvnc; consider adding a Privileged=false property under the X11 section that would allow people to revert to privileged X11

**Comment by Sunderland93** [comment]
Any progress?

**Comment by sandy-8925** [comment]
Is there any remaining work that needs to be done for this PR? Are you guys planning to merge this in at any time?

**Comment by steve95** [comment]
> Is there any remaining work that needs to be done for this PR? Are you guys planning to merge this in at any time?

I really hope so! Running the login manager and consequently the entire desktop environment without root privileges should not be considered a secondary, optional thing, expecially considering the state of Xorg security.

**Comment by Ryujinra** [comment]
Is there any plan going forward to merge this PR at this point? Are there any unmentioned problems blocking this?

**Comment by davidedmundson** [comment]
There are mentioned problems.

**Comment by Ryujinra** [comment]
> There are mentioned problems.

@plfiorini made several commits in Feb 2017 which added the new greeter/server structure and appear to address the issues raised in Oct 2016 regarding the ability to continue to run a privileged xorg instance. Evidently, wayland user sessions have not been tested but if the conflicts can be resolved we can start testing the untested cases described in #673 91a8143c51394ae163ba8096f9b28407a3c9520b.

**Comment by ariasuni** [comment]
Any news on this?

**Comment by devurandom** [comment]
@plfiorini You closed this, but according to GitHub it was not merged. What is the plan for this and #246?

**Comment by Myrddin-Wyllt** [comment]
@devurandom Looks like he's working on it in the [xorguser2019 branch.](https://github.com/plfiorini/sddm/tree/feature/xorguser2019)

**Comment by devurandom** [comment]
> @devurandom It looks like he's working on it in the [xorguser2019 branch.](https://github.com/plfiorini/sddm/tree/feature/xorguser2019)

Thank you, @Myrddin-Wyllt, for figuring that out. I do not see any PR for this and the branch is still live. Any idea when this might be coming?

**Pull request closed without merging**